Refactor ReaderModificationCommandBatch

roji opened the pull request:
The previous implementation of ReaderModificationCommandBatch was written around the assumption that the number of affected rows is fetched via an additional SQL query appended to the batch. This isn't for PostgreSQL, where the affected rows are only returned in the protocol and exposed via a special Npgsql API.

Closes #1956

dnfclas commented:
__@roji__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

AndriySvyryd commented on src/EntityFramework.Relational/Update/AffectedCountModificationCommandBatch.cs:
```diff
@@ -0,0 +1,313 @@
+            return added;
+        }
+
+        public override int Execute(
```
The code outside "using reader" has no assumptions about the reader contents so it can be pulled up.

AndriySvyryd commented on src/EntityFramework.Relational/Update/AffectedCountModificationCommandBatch.cs:
```diff
@@ -0,0 +1,313 @@
+            return commandIndex;
+        }
+
+        private int ConsumeResultSetWithPropagation(int commandIndex, DbDataReader reader, DbContext context)
```
The Consume* methods should be protected virtual

AndriySvyryd commented on src/EntityFramework.Relational/Update/AffectedCountModificationCommandBatch.cs:
```diff
@@ -0,0 +1,313 @@
+                        expectedRowsAffected++;
+                    }
+
+                    throw new DbUpdateConcurrencyException(
```
The code for throwing this exception is pretty generic, so it can be extracted as a method and pulled up

AndriySvyryd commented:
You should also make SqlGenerator.AppendSelectAffectedCountCommand virtual

AndriySvyryd commented on src/EntityFramework.Relational/Update/ReaderModificationCommandBatch.cs:
```diff
@@ -248,181 +221,16 @@ protected virtual void UpdateCachedCommandText(int commandPosition)
-                    {
-                        expectedRowsAffected++;
-                    }
+        protected abstract void Execute(DbDataReader reader, DbContext context);
```
Since the command has already been executed consider renaming this to `Consume`

AndriySvyryd commented:
@roji please also update  `EntityFramework.Relational.Tests\UpdateReaderModificationCommandBatchTest.cs` and run `build.cmd`

roji commented:
Thanks for your patience @AndriySvyryd, I'm a newbie contributing changes to EF7.

I've fixed `EntityFramework.Relational.Tests\UpdateReaderModificationCommandBatchTest.cs` to inherit `AffectedCountModificationCommandBatch` and it now compiles fine with dnx451. However, net46 fails with the following exception:
```
C:\projects\npgsql\external\EntityFramework\test\EntityFramework.Relational.Tests\Migrations\Infrastructure\MigrationIdGeneratorTest.cs(17,13): error CS0012: The type 'Regex' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Text.RegularExpressions, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
```

Running build.cmd (I haven't been doing this till now, just `dnu compile` on specific projects) fails on dnxcore50 on EntityFramework.Commands since System.IServiceProvider doesn't seem to exist.

I guess something's wrong with the way I've set up my environment, but these errors at least seem unrelated to this PR. Please let me know if you'd like any further fixes.

AndriySvyryd commented:
@roji Sounds like you are using an older dnx, try running `packages\KoreBuild\build\dnvm.cmd upgrade` from the enlistment root. The current version is beta6-12120
After upgrading run 'git clean -xdf' to start from a clean state.

As for the PR you also need to update `test\EntityFramework.Relational.Tests\Update\BatchExecutorTest.cs`, the mocks for `ModificationCommandBatch` need to return a non-null value from `.ModificationCommands`

roji commented:
Fixed the mocks, thanks again for the patience. Rebased on latest dev and push-forced.

As for the environment, I'm on 1.0.0-beta6-12120 (clr, x86) and am getting the build errors [above](https://github.com/aspnet/EntityFramework/pull/2448#issuecomment-115374259).

AndriySvyryd commented:
Now `Keys_generated_in_batches` in `test\EntityFramework.SqlServer.FunctionalTests\CommandConfigurationTest.cs` and `Batches_are_divided_correctly_with_two_inserted_columns` in `test\EntityFramework.SqlServer.FunctionalTests\BatchingTest.cs` are failing.

You can also run the tests using http://testdriven.net/

roji commented:
Fixed and rebased on latest dev.

Test fail list for EntityFramework.Relational.Tests, EntityFramework.SqlServer.FunctionalTests looks OK.

Do you guys happen to have a publicly-accessible build server where I can track see the build status of my PRs, failing tests etc?

AndriySvyryd commented:
We do not have a publicly visible CI server, since everyone with commit rights have internal access. But the are some investigations going on that would make it easier to make sure the PR passes the tests.

roji commented:
OK, thanks again for your patience!
